File: dynamic_listings/python/default/developing_mechanics/inertia.py
# ...
class Player:
    input_accel: Vector2 = Vector2()  # Defines the direction we are accelerating
    velocity: Vector2 = Vector2()  # Defines the direction and magnitude of our speed
    position: Vector2 = Vector2()  # Defines our current position, in (x,y) coordinates
    is_moving: bool = False  # Tells us if we're moving
    MAX_SPEED: float = 50.0  # Maximum speed
    ACCEL: float = 15.0  # The acceleration rate
    DECEL: float = 30.0  # The deceleration rate

    def handle_input(self) -> None:
        # First of all, we need to zero the input_accel,
        # or we'll be working on "residual data"
        self.input_accel = Vector2.ZERO
        # Now we can handle movement
        if KEYBOARD.is_left_arrow_pressed:
            self.input_accel.x = self.input_accel.x - 1
        if KEYBOARD.is_right_arrow_pressed:
            self.input_accel.x = self.input_accel.x + 1
        if KEYBOARD.is_down_arrow_pressed:
            self.input_accel.y = self.input_accel.y + 1
        if KEYBOARD.is_up_arrow_pressed:
            self.input_accel.y = self.input_accel.y - 1
        # If any component of the acceleration vector isn't zero, we are moving
        if self.input_accel != Vector2.ZERO:
            self.is_moving = True

    def handle_movement(self, dt: float) -> None:
        if self.is_moving:
            # Vectors will take care of summing forces
            self.velocity = self.velocity + self.ACCEL * dt * self.input_accel
            # We need to clamp the speed, to avoid going too fast
            self.velocity.clamp(MAX_SPEED)
        else:
            # We are stopping, let's subtract the deceleration
            velocity_value: float = len(self.velocity) - self.DECEL * dt
            if velocity_value < 0:
                # If, After decelerating, we have a negative value, we need
                # to make it zero or the object will start moving backwards
                velocity_value = 0
            # We are just changing the length of the vector, so we can just
            # clamp its length
            self.velocity.clamp(velocity_value)

        # Now it's time to move the object
        self.position = self.position + self.velocity

File: dynamic_listings/rust/default/developing_mechanics/inertia.rs
// ...
struct Player {
    input_accel: Vector2D,
    velocity: Vector2D,
    position: Vector2D,
    is_moving: bool,
}

impl Player {
    const MAX_SPEED: f32 = 50.;
    const ACCEL: f32 = 15.;
    const DECEL: f32 = 30.;

    pub fn handle_input(&mut self) {
        // First of all, we need to zero the input_accel, or we'll be working on "residual data"
        self.input_accel = Vector2D::default();
        // Now we can handle movement
        if KEYBOARD.left_arrow_pressed {
            self.input_accel.x -= 1.;
        }
        if KEYBOARD.right_arrow_pressed {
            self.input_accel.x += 1.;
        }
        if KEYBOARD.down_arrow_pressed {
            self.input_accel.y += 1.;
        }
        if KEYBOARD.up_arrow_pressed {
            self.input_accel.y -= 1.;
        }
        // If any component of the acceleration vector is not zero, we are moving
        if self.input_accel != Vector2D::default() {
            self.is_moving = true;
        }
    }

    pub fn handle_movement(&mut self, dt: f32) {
        if self.is_moving {
            // Vectors will take care of summing forces
            self.velocity += self.input_accel * Self::ACCEL * dt;
            // We need to clamp the speed, to avoid going too fast
            self.velocity.max_each(Self::MAX_SPEED);
        } else {
            // We are stopping, let's subtract the deceleration
            let mut velocity_value = self.velocity.length() - Self::DECEL * dt;
            if velocity_value < 0. {
                // If, After decelerating, we have a negative value, we need to make it zero or the object will start moving backwards
                velocity_value = 0.;
            }
            // We are just changing the length of the vector, so we can just clamp its length
            self.velocity.max_each(velocity_value);
        }
        // Now it's time to move the object
        self.position += self.velocity;
    }
}
